File: src/python/portfolio_management/online/cursor_manager.py
"""Cursor Management for Online Programs - migrated from CURSMGR.cbl.

Manages cursor declarations and lifecycle, implements cursor optimization,
handles array fetching for performance, provides cursor status monitoring.
"""

import logging
from dataclasses import dataclass, field
from typing import Optional, Protocol

from portfolio_management.models.common import ReturnCode

logger = logging.getLogger(__name__)

PROGRAM_ID = "CURSMGR"
DEFAULT_MAX_ROWS = 20


class DatabaseCursor(Protocol):
    def execute(self, sql: str, params: Optional[tuple] = None) -> None: ...
    def fetchone(self) -> Optional[tuple]: ...
    def fetchmany(self, size: int) -> list[tuple]: ...
    def close(self) -> None: ...


@dataclass
class CursorInfo:
    name: str = ""
    sql: str = ""
    status: str = "CLOSED"
    rows_fetched: int = 0
    cursor: Optional[DatabaseCursor] = field(default=None, repr=False)


class CursorManager:
    def __init__(self, max_rows: int = DEFAULT_MAX_ROWS):
        self._cursors: dict[str, CursorInfo] = {}
        self._max_rows = max_rows

    def declare(self, name: str, sql: str) -> int:
        if name in self._cursors:
            logger.warning("Cursor %s already declared, replacing", name)

        self._cursors[name] = CursorInfo(name=name, sql=sql, status="DECLARED")
        logger.debug("Cursor %s declared", name)
        return ReturnCode.SUCCESS

    def open(self, name: str, cursor: Optional[DatabaseCursor] = None) -> int:
        info = self._cursors.get(name)
        if info is None:
            logger.error("Cursor %s not declared", name)
            return ReturnCode.ERROR

        info.cursor = cursor
        info.status = "OPEN"
        info.rows_fetched = 0

        if cursor is not None:
            try:
                cursor.execute(info.sql)
            except Exception as e:
                logger.error("Error opening cursor %s: %s", name, e)
                info.status = "ERROR"
                return ReturnCode.ERROR

        logger.debug("Cursor %s opened", name)
        return ReturnCode.SUCCESS

    def fetch(self, name: str, max_rows: Optional[int] = None) -> tuple[int, list]:
        info = self._cursors.get(name)
        if info is None or info.status != "OPEN":
            logger.error("Cursor %s not open", name)
            return ReturnCode.ERROR, []

        fetch_size = max_rows if max_rows is not None else self._max_rows

        if info.cursor is not None:
            try:
                rows = info.cursor.fetchmany(fetch_size)
                info.rows_fetched += len(rows)
                if not rows:
                    return ReturnCode.WARNING, []
                return ReturnCode.SUCCESS, rows
            except Exception as e:
                logger.error("Error fetching from cursor %s: %s", name, e)
                return ReturnCode.ERROR, []

        return ReturnCode.WARNING, []

    def close(self, name: str) -> int:
        info = self._cursors.get(name)
        if info is None:
            logger.warning("Cursor %s not found", name)
            return ReturnCode.WARNING

        if info.cursor is not None:
            try:
                info.cursor.close()
            except Exception as e:
                logger.error("Error closing cursor %s: %s", name, e)

        info.status = "CLOSED"
        info.cursor = None
        logger.debug("Cursor %s closed (fetched %d rows)", name, info.rows_fetched)
        return ReturnCode.SUCCESS

    def get_status(self, name: str) -> Optional[str]:
        info = self._cursors.get(name)
        return info.status if info is not None else None

    def close_all(self) -> int:
        for name in list(self._cursors.keys()):
            self.close(name)
        return ReturnCode.SUCCESS

File: ts-src/programs/online/cursor-manager.ts
/**
 * Cursor Manager.
 * Migrated from: src/programs/online/CURSMGR.cbl
 *
 * Replaces DB2 DECLARE/OPEN/FETCH/CLOSE cursor operations with
 * a query-result-set manager using async iteration and pagination.
 */

import { Knex } from 'knex';
import { ReturnCode } from '../../types';

/** Represents an open cursor (query result set). */
interface CursorState {
  name: string;
  rows: Record<string, unknown>[];
  index: number;
  isOpen: boolean;
}

export class CursorManager {
  private cursors: Map<string, CursorState> = new Map();

  constructor(private readonly db: Knex) {}

  /** DECLARE – register a cursor name and its SQL statement. */
  declare(name: string): number {
    if (this.cursors.has(name)) {
      return ReturnCode.Warning; // already declared
    }
    this.cursors.set(name, { name, rows: [], index: 0, isOpen: false });
    return ReturnCode.Success;
  }

  /** OPEN – execute the query and populate the result set. */
  async open(name: string, query: Knex.QueryBuilder): Promise<number> {
    const cursor = this.cursors.get(name);
    if (!cursor) {
      console.error(`Cursor ${name} not declared`);
      return ReturnCode.Error;
    }

    try {
      cursor.rows = await query;
      cursor.index = 0;
      cursor.isOpen = true;
      return ReturnCode.Success;
    } catch (err) {
      console.error(`Error opening cursor ${name}: ${err}`);
      return ReturnCode.Error;
    }
  }

  /** FETCH – return the next row, or undefined at EOF. */
  fetch(name: string): { rc: number; row?: Record<string, unknown> } {
    const cursor = this.cursors.get(name);
    if (!cursor || !cursor.isOpen) {
      return { rc: ReturnCode.Error };
    }

    if (cursor.index >= cursor.rows.length) {
      return { rc: ReturnCode.Warning }; // EOF (SQLCODE +100)
    }

    const row = cursor.rows[cursor.index++];
    return { rc: ReturnCode.Success, row };
  }

  /** CLOSE – release the result set. */
  close(name: string): number {
    const cursor = this.cursors.get(name);
    if (!cursor) {
      return ReturnCode.Warning;
    }

    cursor.rows = [];
    cursor.index = 0;
    cursor.isOpen = false;
    return ReturnCode.Success;
  }

  /** Close all open cursors. */
  closeAll(): void {
    for (const [name] of this.cursors) {
      this.close(name);
    }
  }
}
